data/modules/MissionUtils/ShipBuilder.lua

```diff
@@ -0,0 +1,855 @@
+
+-- || Hull Threat Factor ||
+
+-- A ship hull starts with at least this much threat
+ShipBuilder.kBaseHullThreatFactor = 1.0
+
+-- How much hull health contributes to ship threat factor (tons of armor to threat)
+ShipBuilder.kArmorToThreatFactor = 0.15
+
+-- How much the ship's maximum forward acceleration contributes to threat factor
+ShipBuilder.kAccelToThreat = 0.02
+-- Tbreat from acceleration is added to this number to determine the final modifier for ship hull HP
+ShipBuilder.kAccelThreatBase = 0.5
+
+-- Controls how a ship's atmospheric performance contributes to its threat factor
+ShipBuilder.kAeroStabilityToThreat = 0.20
+ShipBuilder.kAeroStabilityThreatBase = 0.80
+
+ShipBuilder.kCrossSectionToThreat = 1.0
+ShipBuilder.kCrossSectionThreatBase = 0.75
+
+-- Only accept ships where the hull is at most this fraction of the desired total threat factor
+ShipBuilder.kMaxHullThreatFactor = 0.8
+ShipBuilder.kMinHullThreatFactor = 0.12
+
+-- || Weapon Threat Factor ||
+
+-- Damage in unit tons of armor per second to threat factor
+ShipBuilder.kWeaponDPSToThreat = 0.35
+
+-- Weapon projectile speed modifies threat, calculated as 1.0 + (speed / 1km/s) * mod
+ShipBuilder.kWeaponSpeedToThreat = 0.2
+
+-- Threat factor increase for dual-fire weapons
+ShipBuilder.kDualFireThreatFactor = 1.2
+
+-- Threat factor increase for beam lasers
+ShipBuilder.kBeamLaserThreatFactor = 1.8
+
+-- Amount of hull threat included in this weapon's threat per unit of weapon damage
+ShipBuilder.kWeaponSharedHullThreat = 0.02
+
+-- || Shield Threat Factor ||
+
+-- Threat factor per unit armor ton equivalent of shield health
+-- (compare with hull threat factor per unit ton)
+ShipBuilder.kShieldThreatFactor = 0.2
+
+-- Portion of hull threat factor to contributed per unit ton
+ShipBuilder.kShieldSharedHullThreat = 0.005
+
+-- =============================================================================
+
+-- Class: MissionUtils.ShipTemplate
+--
+-- This class is the primary data container for mission modules to express how
+-- they would like spawned ships to be outfitted.
+--
+-- Pre-existing templates can be found in MissionUtils.ShipTemplates, and
+-- individual mission modules can construct their own specialized templates or
+-- clone and modify existing templates.
+--
+-- Note that the list of equipment rules is shared between clones of a template
+-- for performance reasons and should be overwritten in cloned templates with a
+-- modified copy as desired.
+--
+---@class MissionUtils.ShipTemplate
+---@field clone fun(self, mixin: { rules: MissionUtils.OutfitRule[] }): self
+local Template = utils.proto("MissionUtils.ShipTemplate")
+
+Template.role = nil ---@type string?
+Template.hyperclass = nil ---@type number?
+Template.shipId = nil ---@type string?
+Template.label = nil ---@type string?
+Template.randomModifier = nil ---@type number? Scalar modifying the randomChance value of contained equipment rules
+Template.rules = {} ---@type MissionUtils.OutfitRule[]
+
```

Comment:
Suggestion: add Hull as a template parameter if someone wants a specific ship?

(See related comment in Assassination.lua - https://github.com/pioneerspacesim/pioneer/pull/5734/files#r1827285980)

Reply:
Once I figure out how HullConfigs are going to work (specifically, do they share a ship.json, are they separate shipdefs, etc.) the shipId parameter will be converted to a HullConfig identifier.